eessi/testsuite/hooks.py

```diff
@@ -299,11 +299,15 @@ def filter_valid_systems_by_device_type(test: rfm.RegressionTest, required_devic
         # GPU_VENDOR=GPU_VENDORS[NVIDIA] extras
         valid_systems = f'+{FEATURES[GPU]} %{GPU_VENDOR}={GPU_VENDORS[NVIDIA]}'
 
-    elif required_device_type == DEVICE_TYPES[CPU]:
+    elif not is_cuda_module and required_device_type == DEVICE_TYPES[CPU]:
         # Using the CPU requires partitions with FEATURES[CPU] feature
         # Note: making FEATURES[CPU] an explicit feature allows e.g. skipping CPU-based tests on GPU partitions
         valid_systems = f'+{FEATURES[CPU]}'
 
+    elif is_cuda_module and required_device_type == DEVICE_TYPES[CPU]:
```

Comment:
Essentially, here, you are saying:

"A test requesting a CPU but for a module compiled with CUDA support should only be run if the partition offers both the `cpu` and `gpu` features".

We should discuss if we want this done at the hook level, which would make this the default behavior for all tests. In principle, there are three types of codes:

1. Codes that `dlopen` the CUDA libraries when needed. These codes don't even need the CUDA stubs libraries to be present, as the code path should never encounter this `dlopen` call on a CPU-only node.
2. Codes that are dynamically linked against the CUDA libraries, but do not do any CUDA function calls (e.g. because you tell them explicitly to execute on CPU). These can run, as long as the stubs libraries are found (to satisfy the runtime linker).
3. Codes that are dynamically linked against the CUDA libraries, and that (always) execute CUDA function calls. These _can not_ run on a CPU only node.

Although initially I really disliked setting the `LD_LIBRARY_PATH`, I'm starting to change my mind on that. Both (1) and (2) are a valid thing to do. We probably have non-GPU builds for _most_ modules, but maybe not all (or maybe not all versions). That might cause some users to use a CUDA-build on a CPU only node. That, in my view, is completely legitimate. The user would also have to make the `stubs` libs findable (by setting LD_LIBRARY_PATH), so its not completely trivial, but ok.

One thing we might wonder is if it is _worth_ to do all this extra testing. It would mean testing of all CUDA builds for GROMACS on CPU nodes, likewise for OSU. That's a fair bit of extra testing, for a very limited use case (how many people will actually do this?).

Maybe we should just go for the clean option, which is the hook as you propose it here (because it is clean, and explainable). It might mean some niche software useage is not tested, but we can at least be very clear about that.

Let's discuss in tomorrows meeting.

Reply:
this is exactly what we already discussed in last meeting :)